teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+  syntax support
+* Adding support for nesting, e.g. array params where the items in the array are themselves arrays
+  (no reason we can't add that later but trying to keep it simple for now)
+
+### Use Cases
+
+1. Looping (i.e. providing values for an interface such as the task-loop custom task
+   [iterateParam](https://github.com/tektoncd/experimental/tree/main/task-loops#specifying-the-iteration-parameter)
+   (see the workaround added in [experimental#713](https://github.com/tektoncd/experimental/pull/713) which
+   splits a string). For example, in tektoncd/plumbing we have a directory of
+   [images](https://github.com/tektoncd/plumbing/tree/main/tekton/images). Each directory contains a Dockerfile -
+   in a Pipeline that builds all images in this repo, we would want to: 1) determine which Dockerfiles need to be built
+   and 2) for each of those Dockerfiles, build the image. Looping support would allow us to use a Task which builds
+   just one image (e.g. [the kaniko task](https://github.com/tektoncd/catalog/tree/main/task/kaniko/0.4)), and run it
+   for each Dockerfile, instead of having to the task to take multiple Dockerfiles as input. This missing piece of this
+   is supporting the Task which determines with Dockerfiles to build and provides them as an array (instead of
+   encoded in a string).
+
+   ```
+   filter-dockerfiles
+        |
+        | result: list of dockerfiles
+        v
+   (for each: filter-dockerfiles result)
+      kaniko build
+   ```
+
+## Requirements
+
+* Must be possible for a result array to be empty
+* Must be possible to use the array results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has an array param
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+Currently we treat the contents of
+[a Task's result file(s)](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+as a string. The file is located at `$(results.resultName.path)` which expands to `/tekton/results/resultName`.
+In this proposal, we add the option to write to a json file instead, available at `$(results.resultName.jsonPath)`
+which expands to `/tekton/results/resultName.json`. The contents of this file would be treated as json and in addition
+to supporting a string value in this file, we sould support json arrays.
+
+This adds some additional complexity in that there are two paths to which you could write results (the string file
+and the json file), but this option has these advantages:
+
+* Completely backward compatible with current use of the results file
```

Comment:
Can we not just use the same path and under-the-hood always use json with a `format` key telling us if the content is meant to be read out as a string, dict, or array?

Reply:
that is a GREAT idea! :D